src/replication.c

```diff
@@ -197,6 +204,75 @@ void rebaseReplicationBuffer(long long base_repl_offset) {
     }
 }
 
+/* Replication: Master side - connections association.
+ * On rdb-channel sync, connection association is used to keep replication data in 
+ * the backlog until the replica requests PSYNC. Association happens in two forms, 
+ * if there's an existing buffer block at the fork time, the replica is attached
+ * to the tail, if there is no tail, the replica will be attached when a new
+ * buffer block is created (see the Retrospect function below).
+ * Replica rdb client id is used as a unique key for the association. 
+ * On COB overrun, association is deleted and the RDB connection 
+ * is dropped.
+ */
+void addSlaveToPsyncWaitingDict(client* slave) {
+    listNode *ln = NULL;
+    replBufBlock *tail = NULL;
+    if (server.repl_backlog == NULL) {
+        createReplicationBacklog();
+    } else {
+        ln = listLast(server.repl_buffer_blocks);
+        tail = ln ? listNodeValue(ln) : NULL;
+        if (tail) {
+            tail->refcount++;
+        }
+    }
+    serverLog(LL_DEBUG, "Peer slave %s with cid %ld, %s ", replicationGetSlaveName(slave), slave->id,
+        tail? "with repl-backlog tail": "repl-backlog is empty");
+    slave->ref_repl_buf_node = tail? ln: NULL;
+    /* Prevent rdb client from being freed before psync is established. */
+    slave->flags |= CLIENT_PROTECTED;
+    dictAdd(server.slaves_waiting_psync, (void*)slave->id, slave);
```

Comment:
It will raise a bug on 32bit machine! As @oranagra suggested we should use rax instead of the dictionary as it is already implemented in this context. Thanks!